src/indexing.jl

```diff
@@ -114,13 +114,22 @@ end
 @propagate_inbounds Base.setindex!(A::AxisArray, v, idx::AbstractArray{Bool}) = (A.data[idx] = v)
 
 ### Fancier indexing capabilities provided only by AxisArrays ###
-@propagate_inbounds Base.getindex(A::AxisArray, idxs...) = A[to_index(A,idxs...)...]
-@propagate_inbounds Base.setindex!(A::AxisArray, v, idxs...) = (A[to_index(A,idxs...)...] = v)
+# To avoid StackOverflowErrors on indexes that we don't know how to convert, we
+# give AxisArrays once chance to convert into known format and then defer to the parent
+@propagate_inbounds Base.getindex(A::AxisArray, idxs...) = getindex_converted(A, to_index(A,idxs...)...)
+@propagate_inbounds Base.setindex!(A::AxisArray, v, idxs...) = setindex!_converted(A, v, to_index(A,idxs...)...)
 # Deal with lots of ambiguities here
 @propagate_inbounds Base.view(A::AxisArray, idxs::ViewIndex...) = view(A, to_index(A,idxs...)...)
 @propagate_inbounds Base.view(A::AxisArray, idxs::Union{ViewIndex,AbstractCartesianIndex}...) = view(A, to_index(A,Base.IteratorsMD.flatten(idxs)...)...)
 @propagate_inbounds Base.view(A::AxisArray, idxs...) = view(A, to_index(A,idxs...)...)
 
+@propagate_inbounds getindex_converted(A, idxs::Idx...) = A[idxs...]
+@propagate_inbounds setindex!_converted(A, v, idxs::Idx...) = (A[idxs...] = v)
+
+@propagate_inbounds getindex_converted(A, idxs...) = A.data[idxs...]
```

Comment:
Yep. If AxisArrays doesn't know how to handle the index, then I don't think it can make sensible decisions about whether to return the axis wrapper or not. The particular case I was worried about returns a scalar, so of course it's biased to do that. If it should return more then I think one needs some glue code (i.e., via Requires or similar).